### ISSUE TITLE
speed up preprocessing by rearranging the metatensor transforms

### ISSUE THREAD

**Issue by wyli** [Project-MONAI/tutorials#842]
**Is your feature request related to a problem? Please describe.**
the current tutorials such as [`fast_training_tutorial`](https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb) uses the following preprocessing sequences:

```
LoadImage
EnsureChannelFirst
Orientation
Spacing
ScaleIntensityRange
CropForeground
```

where the `Spacing` transform is often a performance bottleneck. If that's the case, it might be helpful to do the `Spacing` after cropping (see the example sequence below). This won't alter the preprocessing outcomes -- with the metatensor the cropped patches also have the correct spacing/orientation information and ready for further spacing/orientation operations.

```
LoadImage
EnsureChannelFirst
ScaleIntensityRange
CropForeground
Orientation
Spacing
```

Would be great to revise the relevant tutorials and look for performance enhancements by rearranging the transforms.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I would suggest the below sequence as both `crop` and `spacing` reduce the image size in these tutorials, we should do them as soon as possible:
```
LoadImage
EnsureChannelFirst
CropForeground
Orientation
Spacing
ScaleIntensityRange
```
What do you think?

Thanks.

**Comment by wyli** [comment]
Hi @Nic-Ma this ticket is a feature request for more investigations on the sequence, the final optimal solution should be determined on the actual outputs of the tutorials.

**Comment by KumoLiu** [comment]
Hi @wyli @Nic-Ma,
I just had a quick test on [fast_training_tutorial](https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb) with 6 epochs.
Regular PyTorch training results:
1st: Orientation/Spacing/ScaleIntensityRange/CropForeground (current sequence)
`total time of 6 epochs with regular PyTorch training: 213.3626`
2nd: ScaleIntensityRange/CropForeground/Orientation/Spacing
`total time of 6 epochs with regular PyTorch training: 220.0144`
3rd: CropForeground/Orientation/Spacing/ScaleIntensityRange
`total time of 6 epochs with regular PyTorch training: 211.3136`

MONAI fast training results:
1st: Orientation/Spacing/ScaleIntensityRange/CropForeground (current sequence)
`total time of 6 epochs with MONAI fast training: 8.8327, time of preparing cache: 31.4784`
2nd: ScaleIntensityRange/CropForeground/Orientation/Spacing
`total time of 6 epochs with MONAI fast training: 8.6400, time of preparing cache: 29.1753`
3rd: CropForeground/Orientation/Spacing/ScaleIntensityRange
`total time of 6 epochs with MONAI fast training: 8.5824, time of preparing cache: 28.6025`

**Comment by wyli** [comment]
thanks, do they achieve the same convergence speed? would be great to look into the relevant `dtype` option for the transforms.

**Comment by KumoLiu** [comment]
ok， I will try more epochs to see whether the convergence speed are different.

**Comment by Nic-Ma** [comment]
Hi @KumoLiu ,

I think no need to test `ScaleIntensityRange/CropForeground/Orientation/Spacing` anymore, it's definitely the slowest sequence.

Thanks.

**Comment by wyli** [comment]
sure, the foreground can sometimes depend on the intensity values though: `range_func("CropForeground", CropForegroundd(keys=["image", "label"], source_key="image"))`

**Comment by Nic-Ma** [comment]
Hi @wyli ,

OK, so in theory, we should do `ScaleIntensityRange` before `CropForeground` to identify the "background", and `CropForeground/Orientation/Spacing/ScaleIntensityRange` is actually not correct?
But 2nd is slower than the current 1st, we don't need to change anything?

Thanks.

**Comment by KumoLiu** [comment]
I have a little confusion, why `CropForeground` in this tutorial according to the image to crop, normally we use label as source_key.

**Comment by wyli** [comment]
Hi @Nic-Ma, shall we carefully review the transforms, get some proper results (convergence speed, total time, memory), then draw conclusions? I think with the metatensor, there are some possible new opportunities, we should revisit the tutorials and our best practices.

**Comment by wyli** [comment]
> I have a little confusion, why `CropForeground` in this tutorial according to the image to crop, normally we use label as source_key.

that's mainly to make the preprocssing not depend on the label at inference time

**Comment by KumoLiu** [comment]
Hi @wyli @Nic-Ma,
I have tried several experiments on [fast_training_tutorial](https://github.com/Project-MONAI/tutorials/blob/main/acceleration/fast_training_tutorial.ipynb) with 400 epochs. And in view of your discussion yesterday, I did not carry out the experiment with  `ScaleIntensityRange` behind `CropForeground`. Results are shown below. 

1st: Orientation/Spacing/ScaleIntensityRange/CropForeground (current sequence)
regular PyTorch training, convergence from around 200 epochs
```
train completed, best_metric: 0.9487 at epoch: 395 total time: 19019.7932
total time of 400 epochs with regular PyTorch training: 19020.2692
```
MONAI fast training, convergence from around 75 epochs
```
train completed, best_metric: 0.9532 at epoch: 380 total time: 509.6325
total time of 400 epochs with MONAI fast training: 509.6325, time of preparing cache: 33.0693
```
2nd: ScaleIntensityRange/CropForeground/Orientation/Spacing (dtype=float32 in ScaleIntensityRange)
regular PyTorch training, convergence from around 250 epochs
```
train completed, best_metric: 0.9481 at epoch: 400 total time: 21004.5841
total time of 400 epochs with regular PyTorch training: 21004.8646
```
MONAI fast training, convergence from around 50 epochs
```
train completed, best_metric: 0.9555 at epoch: 395 total time: 506.4334
total time of 400 epochs with MONAI fast training: 506.4334, time of preparing cache: 29.0069
```
3rd: ScaleIntensityRange/CropForeground/Orientation/Spacing (dtype=float16 in ScaleIntensityRange)
regular PyTorch training, convergence from around 300 epochs
```
train completed, best_metric: 0.9470 at epoch: 400 total time: 16873.8214
total time of 400 epochs with regular PyTorch training: 16874.0519
```
MONAI fast training, convergence from around 65 epochs
```
train completed, best_metric: 0.9566 at epoch: 400 total time: 504.1156
total time of 400 epochs with MONAI fast training: 504.1156, time of preparing cache: 27.8358
```
It seems that putting the crop before spacing improves the time of preparing cache, but the convergence and training speed aren't significantly affected.

**Comment by Nic-Ma** [comment]
Thanks for the great experiments.
I think different random operations also affect the convergence epoch.

Thanks.

**Comment by wyli** [comment]
thanks, so I think we should prefer the new transform sequence in some tutorials. and then the tutorials require monai>=0.9.1. users with previous versions of monai would use the previous tags such as  https://github.com/Project-MONAI/tutorials/tree/0.9.0

**Comment by yuchen-xu** [comment]
@wyli Thanks for the FYI - do you want me to open a PR to make those updates?

**Comment by wyli** [comment]
Hi @yuchen-xu it's for your information only, I think @KumoLiu is working on this.
